tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/caches/Cache.scala

```diff
@@ -0,0 +1,153 @@
+package at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.caches
+
+import at.forsyte.apalache.tla.bmcmt.smt.SolverContext
+import at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.Stackable
+
+import scala.collection.immutable.HashMap
+
+/**
+ * An abstract cache that assigns `TargetT`-typed values to `SourceT` typed values, while potentially modifying a
+ * `ContextT`-typed context.
+ *
+ * Example: IntValueCache assigns [[ArenaCell]]s to integers, while modifying a [[PureArena]].
+ *
```

Comment:
Could you please add a short paragraph explaining *why* this class needs to be able to snapshot and roll back the state? That is the most critical part of the behavior iiuc (otherwise we could just use `lazy` for memoization).